Discard overlapped elements in `hint-mode`

heiwiper opened the pull request:
-------

# Description

I have added a function to verify whether an element is overlapped when calling the `follow-hint` command. This function is used to discard the overlapped elements when the user has set the `compute-hints-in-view-port-p` to `t`.

Fixes #2506

# Discussion

The new added function `element-overlapped-p` is inspired by [the algorithm](https://github.com/lusakasa/saka-key/blob/master/src/modes/hints/client/findHints.js#L114) from `saka-key` project. I have noticed a side effect to this algorithm which is that it automatically discards elements that aren't in the view port, this part is already implemented in the function `element-in-view-port-p` which makes the process redundant since both functions are called when checking for hints. After investigating the issue I found out that the javascript function [`elementFromPoint`](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint) returns `null` when the element is outside the view port, which according to the algorithm acts as if the two elements are not related thus one is overlapping the other.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [ ] I've added the new dependencies as:
  - [ ] ASDF dependencies,
  - [ ] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [ ] and Guix dependencies.
- [ ] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [x] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [x] Documentation:
  - [ ] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [ ] I have updated the existing documentation to match my changes.
  - [ ] I have commented my code in hard-to-understand areas.
  - [ ] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [ ] I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - [ ] (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [ ] Compilation and tests:
  - [x] My changes generate no new warnings.
  - [ ] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [ ] New and existing unit tests pass locally with my changes.

heiwiper commented:
I am not sure if all these tasks should be checked because some of them don't seem to be appropriate to this small change. As for the style guidelines I haven't read them yet, I will do in the coming days.

Please let me know which tasks are mandatory for the pull request to be merged.

aadcg commented:
@heiwiper thanks for this PR! 

> I am not sure if all these tasks should be checked because some of them don't seem to be appropriate to this small change. As for the style guidelines I haven't read them yet, I will do in the coming days.

Don't worry too much about it! It's your first contribution so, realistically, no one expects you to be fully aware of all of the project's protocols.

I'll take a deeper look at your changes right now.

aadcg commented:
> I have noticed a side effect to this algorithm which is that it automatically discards elements that aren't in the view port, this part is already implemented in the function element-in-view-port-p which makes the process redundant since both functions are called when checking for hints. After investigating the issue I found out that the javascript function elementFromPoint returns null when the element is outside the view port

This makes sense, but I'm still confused about one detail.

Make the following test. Drop `nyxt/ps:element-in-view-port-p` in `set-hintable-attribute` (since it's redundant).  Then issue the command `follow-hint` from the very bottom of page `https://github.com/atlas-engineer/nyxt`. If you scroll up, you'll see that one hint has been drawn at the beginning of the README section. I can't make sense of it but, honestly, it's not very important.

I think you're right and we should only check for `nyxt/ps:element-overlapped-p`.  But then, I think we should rename this predicate. Any idea?

aadcg commented on source/mode/hint.lisp:
```diff
@@ -159,7 +159,8 @@ For instance, to include images:
       (if in-view-port-p
-          (when (nyxt/ps:element-in-view-port-p element)
+          (when (and (nyxt/ps:element-in-view-port-p element)
+                     (not (nyxt/ps:element-overlapped-p element)))
```
Let's drop `nyxt/ps:element-in-view-port-p` as you suggest, since it's redundant.

Then you can say `(unless (nyxt/ps:element-overlapped-p ...`.

aadcg commented on source/parenscript-macro.lisp:
```diff
@@ -151,3 +151,21 @@
+  `(let* ((rect (chain ,element (get-bounding-client-rect)))
+          (computed-style (chain window (get-computed-style ,element)))
+          (coord-truncation-offset 2)
+          (radius (serapeum:parse-float (chain computed-style border-top-left-radius)))
```
I believe I understand why you wrote it as `serapeum:parse-float` here: the corresponding JS function (parseFloat) that this code compiles to is indeed similar to `serapeum:parse-float`. 

But I believe this makes it actually more confusing. I'd write it as `parse-float`, what do you think?

aadcg commented on source/parenscript-macro.lisp:
```diff
@@ -151,3 +151,21 @@
+         t
+         (progn (loop while (and el (/= el element))
+                      do (setf el (chain el parent-node)))
+                (null el)))))
```
This `if` block could be refactored with `unless`. Notice that in that case, you don't even need `progn`. Makes sense?

heiwiper replied:
It does make sense but I'm not sure about this specific use case. The `if` statement would return `t` when the condition's evaluation result is `t`, unlike the `unless` which would return a `nil` as if the element isn't overlapped.

aartaka replied:
@heiwiper, that does make sense, yes.

aadcg replied:
@heiwiper, indeed, my suggestion would not yield an equivalent behavior!

aadcg commented on source/parenscript-macro.lisp:
```diff
@@ -151,3 +151,21 @@
+     (if (or (>= offset (chain rect width))
+             (>= offset (chain rect height)))
+         t
+         (progn (loop while (and el (/= el element))
```
I believe it would be more Lispy to write `(not (eq el element))` :)

aartaka replied:
[Parenscript only has](https://parenscript.common-lisp.dev/reference.html#section-arithmetic) `eq`, `eql`, `equal`, and `=`, so yes, `(not (eq ...))` works better here.

aadcg commented on source/parenscript-macro.lisp:
```diff
@@ -151,3 +151,21 @@
+(export-always 'element-overlapped-p)
+(defpsmacro element-overlapped-p (element)
+  "Whether ELEMENT is overlapped by another element."
+  `(let* ((rect (chain ,element (get-bounding-client-rect)))
```
I suggest adding a link to the saka-keys algorithm for reference in a comment.

aadcg commented:
@heiwiper amazing work! I've left some comments, thanks.

heiwiper commented:
> Make the following test. Drop `nyxt/ps:element-in-view-port-p` in `set-hintable-attribute` (since it's redundant). Then issue the command `follow-hint` from the very bottom of page `https://github.com/atlas-engineer/nyxt`. If you scroll up, you'll see that one hint has been drawn at the beginning of the README section. I can't make sense of it but, honestly, it's not very important.

You are right about this I just noticed that behaviour, but it's apparently the same before making the changes I have added. Maybe we could open a new issue for that specific behaviour, I would gladly work on it if you think it's not very complex for a new comer.

heiwiper commented:
I have added the changes locally, I will test it and update the pull request a bit later today.

heiwiper commented:
I would like to thank you for reviews, I have just added the changes you have mentioned and marked the reviews as resolved except one in which I added a reply. Let me know if there are any other changes related to that one unresolved review or any other part of the code.

aartaka commented:
@aadcg, can you test locally? My Nyxt misbehaves atm, so I cannot test |_・)

aadcg commented:
> @aadcg, can you test locally? My Nyxt misbehaves atm, so I cannot test
> |_・)

Sure!  Don't worry, I'm going to check this later today.  I'm
prioritizing energy-consuming tasks and leaving the easier ones for
later.

aadcg commented:
@heiwiper I've merged your work via commits 2575c08ca and 5a9fb1d6a!

Thank you again for your help :)
